internal/scheduler/nodedb.go

```diff
@@ -383,45 +533,72 @@ func (nodeDb *NodeDb) dominantResource(req *schedulerobjects.PodRequirements) st
 	return dominantResourceType
 }
 
-// Upsert nodes.
-func (nodeDb *NodeDb) Upsert(nodes []*schedulerobjects.Node) error {
+func (nodeDb *NodeDb) UpsertMany(nodes []*schedulerobjects.Node) error {
 	txn := nodeDb.db.Txn(true)
 	defer txn.Abort()
+	if err := nodeDb.UpsertManyWithTxn(txn, nodes); err != nil {
+		return err
+	}
+	txn.Commit()
+	return nil
+}
+
+func (nodeDb *NodeDb) UpsertManyWithTxn(txn *memdb.Txn, nodes []*schedulerobjects.Node) error {
 	for _, node := range nodes {
+		if err := nodeDb.UpsertWithTxn(txn, node); err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-		// Compute the node type of the node
-		// and update the node with the node accordingly.
-		nodeType := schedulerobjects.NewNodeType(
-			node.GetTaints(),
-			node.GetLabels(),
-			nodeDb.indexedTaints,
-			nodeDb.indexedNodeLabels,
-		)
-		node.NodeTypeId = nodeType.Id
-		node.NodeType = nodeType
+func (nodeDb *NodeDb) Upsert(node *schedulerobjects.Node) error {
+	txn := nodeDb.Txn(true)
+	defer txn.Abort()
+	if err := nodeDb.UpsertWithTxn(txn, node); err != nil {
+		return err
+	}
+	txn.Commit()
+	return nil
+}
 
-		// Record all unique node types.
+func (nodeDb *NodeDb) UpsertWithTxn(txn *memdb.Txn, node *schedulerobjects.Node) error {
+	// Compute the node type of the node
+	// and update the node with the node accordingly.
+	nodeType := schedulerobjects.NewNodeType(
+		node.GetTaints(),
+		node.GetLabels(),
+		nodeDb.indexedTaints,
+		nodeDb.indexedNodeLabels,
+	)
+	node.NodeTypeId = nodeType.Id
+	node.NodeType = nodeType
+
+	// Record all unique node types.
+	nodeDb.mu.Lock()
+	nodeDb.nodeTypes[nodeType.Id] = nodeType
+	nodeDb.mu.Unlock()
+
+	// If this is a new node, increase the overall resource count.
+	// Note that nodeDb.totalResources isn't rolled back on txn abort.
+	if existingNode, err := nodeDb.GetNodeWithTxn(txn, node.Id); err != nil {
+		return err
+	} else if existingNode == nil {
 		nodeDb.mu.Lock()
-		nodeDb.nodeTypes[nodeType.Id] = nodeType
+		nodeDb.totalResources.Add(node.TotalResources)
 		nodeDb.mu.Unlock()
+	}
 
-		// If this is a new node, increase the overall resource count.
-		it, err := nodeDb.db.Txn(false).Get("nodes", "id", node.Id)
-		if err != nil {
-			return errors.WithStack(err)
-		}
-		if it.Next() == nil {
-			nodeDb.mu.Lock()
-			nodeDb.totalResources.Add(node.TotalResources)
-			nodeDb.mu.Unlock()
-		}
-
-		// Add the node to the db.
-		err = txn.Insert("nodes", node)
-		if err != nil {
-			return errors.WithStack(err)
-		}
+	// Add the node to the db.
+	if err := txn.Insert("nodes", node); err != nil {
+		return errors.WithStack(err)
 	}
+
+	// TODO: Doing this at every node insert seems wasteful.
+	nodeDb.mu.Lock()
+	nodeDb.timeOfMostRecentUpsert = time.Now()
```

Comment:
why do we need this- it used to be used to check liveness for the submit check but I've removed the need for that

Reply:
In that case we don't need it anymore. Let me remove it.